City area

MohanadMaraaba opened the pull request:
'Completed' initial work on Area and City portions

MohanadMaraaba committed:
Added Area and City entities
Added AreaRepository & CityRepository
Added AreaController
Connected to MySQL Database
Created Dummy Data through CommandLineRunner

MohanadMaraaba committed:
AreaController changes
Added get method
Added getAll method
Finished the update method
Added error messages

MohanadMaraaba committed:
Added CityController
in City changed Id -> id (mistake)

MohMattour commented on src/main/java/com/example/ecom/controller/CityController.java:
```diff
@@ -0,0 +1,67 @@
+
+    // Get a specific city
+    @GetMapping("/{id}")
+    public City get(@PathVariable Long id){
```
What if id is null?

MohMattour commented on src/main/java/com/example/ecom/controller/CityController.java:
```diff
@@ -0,0 +1,67 @@
+        this.cityRepository = cityRepository;
+    }
+
+    // Get all cities
```
You should write a better documentation Get all cities is not good, it is same as the method name it is useless. remained me tomorrow to show you how to write it.

MohMattour commented on src/main/java/com/example/ecom/controller/CityController.java:
```diff
@@ -0,0 +1,67 @@
+    public City update(@PathVariable Long id, @RequestBody String cityName){
+        Optional<City> optionalCity = cityRepository.findById(id);
+        if(!optionalCity.isPresent()){
+            throw new NullPointerException("Invalid city id.");
```
Wrong msg you check if the city object exists not the id, thats mean if the object exists and the Id null he will pass the check which is wrong.

MohMattour commented on src/main/java/com/example/ecom/controller/CityController.java:
```diff
@@ -0,0 +1,67 @@
+    // Update a specific city
+    @PutMapping("/{id}")
+    public City update(@PathVariable Long id, @RequestBody String cityName){
+        Optional<City> optionalCity = cityRepository.findById(id);
```
Same as before the id  might be null

MohMattour commented on src/main/java/com/example/ecom/controller/CityController.java:
```diff
@@ -0,0 +1,67 @@
+            throw new NullPointerException("Invalid city id.");
+        }
+        City city = optionalCity.get();
+        city.setCityName(cityName);
```
what if the cityname null?

MohMattour commented on src/main/java/com/example/ecom/controller/CityController.java:
```diff
@@ -0,0 +1,67 @@
+
+    // Delete a specific city
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id){
```
same as before null check

MohMattour reviewed (changes requested):
Check the feedback.

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -20,12 +21,28 @@ public AreaController(AreaRepository areaRepository, CityRepository cityReposito
         Optional<City> optionalCity = cityRepository.findById(id);
         if(!optionalCity.isPresent()) {
-            throw new NullPointerException();
+            throw new NullPointerException("Invalid city id.");
```
this Must be a badRequest exception.

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -40,12 +57,15 @@ public void delete(@PathVariable Long id) {
+    // Update an area
     @PutMapping("/{id}")
-    public void update(@PathVariable Long id, @RequestBody Area area){
+    public Area update(@PathVariable Long id, @RequestBody String areaName){
```
I want the request body as Object since I might add something more than a name like inactive flag.

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -40,12 +57,15 @@ public void delete(@PathVariable Long id) {
     @PutMapping("/{id}")
-    public void update(@PathVariable Long id, @RequestBody Area area){
+    public Area update(@PathVariable Long id, @RequestBody String areaName){
         Optional<Area> optionalArea = areaRepository.findById(id);
```
Null check

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -40,12 +57,15 @@ public void delete(@PathVariable Long id) {
+    public Area update(@PathVariable Long id, @RequestBody String areaName){
         Optional<Area> optionalArea = areaRepository.findById(id);
-        if(!optionalArea.isPresent()){
-            throw new NullPointerException();
```
Bad Request exception

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -40,12 +57,15 @@ public void delete(@PathVariable Long id) {
-        if(!optionalArea.isPresent()){
-            throw new NullPointerException();
+        if(!optionalArea.isPresent()) {
+            throw new NullPointerException("Invalid area id.");
```
Bad Request exception

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -40,12 +57,15 @@ public void delete(@PathVariable Long id) {
         }
-
+        Area area = optionalArea.get();
+        area.setAreaName(areaName);
```
null check

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -20,12 +21,28 @@ public AreaController(AreaRepository areaRepository, CityRepository cityReposito
+    // Get an area
+    @GetMapping("/{id}")
+    public Area get(@PathVariable Long id){
+        Optional<Area> optionalArea = areaRepository.findById(id);
```
null check

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -20,12 +21,28 @@ public AreaController(AreaRepository areaRepository, CityRepository cityReposito
+    public Area get(@PathVariable Long id){
+        Optional<Area> optionalArea = areaRepository.findById(id);
+        if(!optionalArea.isPresent()) {
+            throw new NullPointerException("Invalid area id.");
```
Bad Request exception

MohMattour commented on src/main/java/com/example/ecom/model/City.java:
```diff
@@ -0,0 +1,29 @@
+
+    // Area
+    @JsonManagedReference
+    @OneToMany(mappedBy = "city", fetch = FetchType.EAGER)
```
the fetch type should be lazy

MohMattour commented on src/main/java/com/example/ecom/model/City.java:
```diff
@@ -0,0 +1,29 @@
+    private Long Id;
+    @NonNull
+    private String cityName;
+
```
add inactive flag

MohMattour commented on src/main/java/com/example/ecom/model/Area.java:
```diff
@@ -0,0 +1,28 @@
+    private Long id;
+    @NonNull
+    private String areaName;
+
```
add inactive flag

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -0,0 +1,51 @@
+    public void update(@PathVariable Long id, @RequestBody Area area){
+        Optional<Area> optionalArea = areaRepository.findById(id);
+        if(!optionalArea.isPresent()){
+            throw new NullPointerException();
```
bad request exception

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -0,0 +1,51 @@
+
+    // Delete an area
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
```
null check

MohMattour commented on src/main/java/com/example/ecom/controller/AreaController.java:
```diff
@@ -0,0 +1,51 @@
+    // Add an area to a specific city
+    @PostMapping("/{id}")
+    public Area create(@PathVariable Long id, @RequestBody String areaName){
+        Optional<City> optionalCity = cityRepository.findById(id);
```
null check

MohMattour reviewed (changes requested):
you need to add a null checks, also you NEVER use a null pointer exception you use a bad request to send a response to the user with a msg error

MohanadMaraaba committed:
Area & City fixes

Added null checks for ids
Made update and create methods for CityController and AreaController take objects are parameters
Added null checks for the previously mentioned objects
Created exceptions package and added customer exceptions
Updated documentation

MohanadMaraaba committed:
Added auditing

MohanadMaraaba committed:
Enabled Jpa Auditing

MohanadMaraaba committed:
Enabled Jpa Auditing

MohanadMaraaba committed:
Merge pull request #2 from Scoutiano/model

Model

MohanadMaraaba committed:
Fixed issues with update & delete methods in both City and Area Controllers

Delete now sets entity.active to false
Fixed some issues with update

MohanadMaraaba committed:
Update in City controller fixed

MohanadMaraaba committed:
Deleted product related classes that don't belong in this branch

MohanadMaraaba committed:
Merge pull request #3 from Scoutiano/model

Model
